File: app/src/main/java/com/amaury/pointage/v2/engine/FrenchPublicHolidayCalendarV2.kt
package com.amaury.pointage.v2.engine

import java.time.LocalDate

/**
 * Calendrier déterministe des jours fériés légaux français utilisés pour affecter des minutes
 * réellement travaillées. Il ne décide jamais du taux de paie : cette décision reste dans
 * l'arbitrage ACCO/KALI/LEGI.
 *
 * Sources de structure juridique :
 * - Code du travail L3133-1 : onze fêtes légales communes ;
 * - L3134-13 : régime local Moselle/Bas-Rhin/Haut-Rhin ;
 * - L3422-2 : commémoration de l'abolition de l'esclavage dans les territoires concernés.
 *
 * Le 1er mai est identifié séparément car son régime de rémunération relève d'une règle LEGI propre.
 */
object FrenchPublicHolidayCalendarV2 {
    enum class Jurisdiction {
        COMMON_FRANCE,
        ALSACE_MOSELLE,
        GUADELOUPE,
        MARTINIQUE,
        GUYANE,
        MAYOTTE,
        REUNION,
        SAINT_BARTHELEMY,
        SAINT_MARTIN,
        SPECIAL_TERRITORY_UNKNOWN,
        ADDRESS_UNKNOWN
    }

    data class Scope(
        val jurisdiction: Jurisdiction,
        /** False signifie qu'une règle locale peut ajouter un jour impossible à déduire de l'adresse seule. */
        val complete: Boolean,
        val postalCode: String? = null,
        val warning: String? = null
    )

    fun scopeForAddress(address: String): Scope {
        val postal = Regex("(?<!\\d)(\\d{5})(?!\\d)").find(address)?.groupValues?.getOrNull(1)
            ?: return Scope(
                Jurisdiction.ADDRESS_UNKNOWN,
                complete = false,
                warning = "Adresse de l'entreprise insuffisante pour confirmer le calendrier territorial des jours fériés."
            )

        return when {
            postal == "97133" -> Scope(Jurisdiction.SAINT_BARTHELEMY, true, postal)
            postal == "97150" -> Scope(Jurisdiction.SAINT_MARTIN, true, postal)
            postal.startsWith("971") -> Scope(Jurisdiction.GUADELOUPE, true, postal)
            postal.startsWith("972") -> Scope(Jurisdiction.MARTINIQUE, true, postal)
            postal.startsWith("973") -> Scope(Jurisdiction.GUYANE, true, postal)
            postal.startsWith("974") -> Scope(Jurisdiction.REUNION, true, postal)
            postal.startsWith("976") -> Scope(Jurisdiction.MAYOTTE, true, postal)
            postal.startsWith("98") -> Scope(
                Jurisdiction.SPECIAL_TERRITORY_UNKNOWN,
                complete = false,
                postalCode = postal,
                warning = "Territoire à calendrier local : validation juridique des jours fériés requise avant calcul automatique."
            )
            postal.startsWith("57") || postal.startsWith("67") || postal.startsWith("68") -> Scope(
                Jurisdiction.ALSACE_MOSELLE,
                complete = false,
                postalCode = postal,
                warning = "Alsace-Moselle : le Vendredi saint dépend de la commune ; le calendrier ne peut pas être déclaré exhaustif depuis le seul code postal."
            )
            else -> Scope(Jurisdiction.COMMON_FRANCE, true, postal)
        }
    }

    /** Jours fériés pouvant recevoir une règle collective générique, hors 1er mai. */
    fun genericHolidays(year: Int, scope: Scope): Set<LocalDate> {
        require(year in 1900..2200) { "Année hors périmètre du calendrier" }
        val easter = easterSunday(year)
        val values = linkedSetOf(
            LocalDate.of(year, 1, 1),
            easter.plusDays(1),
            LocalDate.of(year, 5, 8),
            easter.plusDays(39),
            easter.plusDays(50),
            LocalDate.of(year, 7, 14),
            LocalDate.of(year, 8, 15),
            LocalDate.of(year, 11, 1),
            LocalDate.of(year, 11, 11),
            LocalDate.of(year, 12, 25)
        )

        when (scope.jurisdiction) {
            Jurisdiction.ALSACE_MOSELLE -> {
                // Le 26 décembre est certain ; le Vendredi saint reste volontairement non ajouté
                // car son application dépend de la commune (L3134-13).
                values += LocalDate.of(year, 12, 26)
            }
            Jurisdiction.GUADELOUPE,
            Jurisdiction.SAINT_MARTIN -> values += LocalDate.of(year, 5, 27)
            Jurisdiction.MARTINIQUE -> values += LocalDate.of(year, 5, 22)
            Jurisdiction.GUYANE -> values += LocalDate.of(year, 6, 10)
            Jurisdiction.MAYOTTE -> values += LocalDate.of(year, 4, 27)
            Jurisdiction.REUNION -> values += LocalDate.of(year, 12, 20)
            Jurisdiction.SAINT_BARTHELEMY -> values += LocalDate.of(year, 10, 9)
            Jurisdiction.COMMON_FRANCE,
            Jurisdiction.SPECIAL_TERRITORY_UNKNOWN,
            Jurisdiction.ADDRESS_UNKNOWN -> Unit
        }
        return values
    }

    fun mayFirst(year: Int): LocalDate = LocalDate.of(year, 5, 1)

    fun isGenericHoliday(date: LocalDate, scope: Scope): Boolean =
        date in genericHolidays(date.year, scope)

    fun isMayFirst(date: LocalDate): Boolean = date.monthValue == 5 && date.dayOfMonth == 1

    /** Algorithme grégorien de Meeus/Jones/Butcher. */
    internal fun easterSunday(year: Int): LocalDate {
        require(year in 1900..2200)
        val a = year % 19
        val b = year / 100
        val c = year % 100
        val d = b / 4
        val e = b % 4
        val f = (b + 8) / 25
        val g = (b - f + 1) / 3
        val h = (19 * a + b - d - g + 15) % 30
        val i = c / 4
        val k = c % 4
        val l = (32 + 2 * e + 2 * i - h - k) % 7
        val m = (a + 11 * h + 22 * l) / 451
        val month = (h + l - 7 * m + 114) / 31
        val day = (h + l - 7 * m + 114) % 31 + 1
        return LocalDate.of(year, month, day)
    }
}

File: ios/HPTravail/HPTravail/SalaryV2/FrenchPublicHolidayCalendarV2.swift
import Foundation

/// Calendrier déterministe des jours fériés légaux français utilisé uniquement pour
/// affecter des minutes réellement travaillées. Il ne décide jamais d'un taux de paie.
///
/// Port iOS de la couche Android existante : le 1er mai reste séparé, les territoires
/// particuliers sont explicites et un périmètre non exhaustif reste fail-closed.
enum FrenchPublicHolidayCalendarV2 {
    enum Jurisdiction: Equatable {
        case commonFrance
        case alsaceMoselle
        case guadeloupe
        case martinique
        case guyane
        case mayotte
        case reunion
        case saintBarthelemy
        case saintMartin
        case specialTerritoryUnknown
        case addressUnknown
    }

    struct Scope: Equatable {
        let jurisdiction: Jurisdiction
        /// false = des jours locaux peuvent exister sans pouvoir être confirmés depuis l'adresse seule.
        let complete: Bool
        let postalCode: String?
        let warning: String?
    }

    static func scopeForAddress(_ address: String) -> Scope {
        guard let postal = postalCode(in: address) else {
            return Scope(
                jurisdiction: .addressUnknown,
                complete: false,
                postalCode: nil,
                warning: "Adresse de l'entreprise insuffisante pour confirmer le calendrier territorial des jours fériés."
            )
        }

        switch postal {
        case "97133":
            return Scope(jurisdiction: .saintBarthelemy, complete: true, postalCode: postal, warning: nil)
        case "97150":
            return Scope(jurisdiction: .saintMartin, complete: true, postalCode: postal, warning: nil)
        default:
            break
        }

        if postal.hasPrefix("971") {
            return Scope(jurisdiction: .guadeloupe, complete: true, postalCode: postal, warning: nil)
        }
        if postal.hasPrefix("972") {
            return Scope(jurisdiction: .martinique, complete: true, postalCode: postal, warning: nil)
        }
        if postal.hasPrefix("973") {
            return Scope(jurisdiction: .guyane, complete: true, postalCode: postal, warning: nil)
        }
        if postal.hasPrefix("974") {
            return Scope(jurisdiction: .reunion, complete: true, postalCode: postal, warning: nil)
        }
        if postal.hasPrefix("976") {
            return Scope(jurisdiction: .mayotte, complete: true, postalCode: postal, warning: nil)
        }
        if postal.hasPrefix("98") {
            return Scope(
                jurisdiction: .specialTerritoryUnknown,
                complete: false,
                postalCode: postal,
                warning: "Territoire à calendrier local : validation juridique des jours fériés requise avant calcul automatique."
            )
        }
        if postal.hasPrefix("57") || postal.hasPrefix("67") || postal.hasPrefix("68") {
            return Scope(
                jurisdiction: .alsaceMoselle,
                complete: false,
                postalCode: postal,
                warning: "Alsace-Moselle : le Vendredi saint dépend de la commune ; le calendrier ne peut pas être déclaré exhaustif depuis le seul code postal."
            )
        }

        return Scope(
            jurisdiction: .commonFrance,
            complete: true,
            postalCode: postal,
            warning: nil
        )
    }

    /// Jours pouvant recevoir une règle collective générique, hors 1er mai.
    static func genericHolidays(
        year: Int,
        scope: Scope
    ) -> Set<PayrollCivilDateV2>? {
        guard (1900...2200).contains(year),
              let easter = easterSunday(year) else {
            return nil
        }

        var values = Set<PayrollCivilDateV2>()
        [
            date(year, 1, 1),
            addingDays(easter, 1),
            date(year, 5, 8),
            addingDays(easter, 39),
            addingDays(easter, 50),
            date(year, 7, 14),
            date(year, 8, 15),
            date(year, 11, 1),
            date(year, 11, 11),
            date(year, 12, 25)
        ].compactMap { $0 }.forEach { values.insert($0) }

        let local: PayrollCivilDateV2?
        switch scope.jurisdiction {
        case .alsaceMoselle:
            local = date(year, 12, 26)
        case .guadeloupe, .saintMartin:
            local = date(year, 5, 27)
        case .martinique:
            local = date(year, 5, 22)
        case .guyane:
            local = date(year, 6, 10)
        case .mayotte:
            local = date(year, 4, 27)
        case .reunion:
            local = date(year, 12, 20)
        case .saintBarthelemy:
            local = date(year, 10, 9)
        case .commonFrance, .specialTerritoryUnknown, .addressUnknown:
            local = nil
        }
        if let local { values.insert(local) }
        return values
    }

    /// Dates potentiellement fériées que le périmètre disponible ne permet pas de trancher.
    static func unresolvedPossibleHolidays(
        year: Int,
        scope: Scope
    ) -> Set<PayrollCivilDateV2>? {
        guard (1900...2200).contains(year),
              let easter = easterSunday(year) else {
            return nil
        }

        switch scope.jurisdiction {
        case .alsaceMoselle:
            guard let goodFriday = addingDays(easter, -2) else { return nil }
            return [goodFriday]
        case .addressUnknown, .specialTerritoryUnknown:
            return Set(
                [
                    addingDays(easter, -2),
                    date(year, 4, 27),
                    date(year, 5, 22),
                    date(year, 5, 27),
                    date(year, 6, 10),
                    date(year, 10, 9),
                    date(year, 12, 20),
                    date(year, 12, 26)
                ].compactMap { $0 }
            )
        default:
            return []
        }
    }

    static func mayFirst(_ year: Int) -> PayrollCivilDateV2? {
        guard (1900...2200).contains(year) else { return nil }
        return date(year, 5, 1)
    }

    static func isGenericHoliday(
        _ value: PayrollCivilDateV2,
        scope: Scope
    ) -> Bool {
        genericHolidays(year: value.year, scope: scope)?.contains(value) == true
    }

    static func isMayFirst(_ value: PayrollCivilDateV2) -> Bool {
        value.month == 5 && value.day == 1
    }

    /// Algorithme grégorien de Meeus/Jones/Butcher, identique au moteur Android.
    static func easterSunday(_ year: Int) -> PayrollCivilDateV2? {
        guard (1900...2200).contains(year) else { return nil }
        let a = year % 19
        let b = year / 100
        let c = year % 100
        let d = b / 4
        let e = b % 4
        let f = (b + 8) / 25
        let g = (b - f + 1) / 3
        let h = (19 * a + b - d - g + 15) % 30
        let i = c / 4
        let k = c % 4
        let l = (32 + 2 * e + 2 * i - h - k) % 7
        let m = (a + 11 * h + 22 * l) / 451
        let month = (h + l - 7 * m + 114) / 31
        let day = (h + l - 7 * m + 114) % 31 + 1
        return date(year, month, day)
    }

    private static func postalCode(in address: String) -> String? {
        let pattern = #"(?<!\d)(\d{5})(?!\d)"#
        guard let range = address.range(of: pattern, options: .regularExpression) else {
            return nil
        }
        return String(address[range])
    }

    private static func date(_ year: Int, _ month: Int, _ day: Int) -> PayrollCivilDateV2? {
        PayrollCivilDateV2(year: year, month: month, day: day)
    }

    private static func addingDays(
        _ value: PayrollCivilDateV2,
        _ offset: Int
    ) -> PayrollCivilDateV2? {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        guard let base = calendar.date(
            from: DateComponents(year: value.year, month: value.month, day: value.day)
        ),
        let result = calendar.date(byAdding: .day, value: offset, to: base) else {
            return nil
        }
        let components = calendar.dateComponents([.year, .month, .day], from: result)
        guard let year = components.year,
              let month = components.month,
              let day = components.day else {
            return nil
        }
        return PayrollCivilDateV2(year: year, month: month, day: day)
    }
}
